Expose BitmapFactory.Options.inPurgeable through RequestCreator.purgeable()

sneuberger-amazon opened the pull request:
When purgeable() is called, we set inPurgeable and inInputShareable to true. We
also decode the bitmap from a byte array instead of a stream, because
inPurgeable only affects the fomer (we only do this for API level <21 because
inPurgeable has no effect for API level 21+).

dnkoutso commented on picasso/src/main/java/com/squareup/picasso/RequestCreator.java:
```diff
@@ -366,6 +366,16 @@ public RequestCreator networkPolicy(NetworkPolicy policy, NetworkPolicy... addit
     return this;
   }
 
+  /** Set inPurgeable and inInputShareable when decoding. This will force the bitmap to be decoded
```
nit: need extra line here.

dnkoutso commented:
checkstyle fails in the build, also need a test for this.

sneuberger-amazon commented:
1. Fixed a bug where bitmap options could be null (I originally wrote this patch for an old version of Picasso where bitmap options was never null).
2. Wrote tests. I don't have a test for BitmapHunter.decodeStream(). The only way I see to test it would be to extract two methods decodeFromArray() and decodeFromStream() and using a spy() to verify which one is called. Let me know if you want this.
3. Fixed style issues. Build now passes.

sneuberger-amazon commented:
Is there any more work I can do on this?

sneuberger-amazon commented:
Friendly ping, @dnkoutso or any other maintainer who can take a look at this.

dnkoutso commented on picasso/src/main/java/com/squareup/picasso/BitmapHunter.java:
```diff
@@ -120,10 +120,12 @@ static Bitmap decodeStream(InputStream stream, Request request) throws IOExcepti
     final boolean calculateSize = RequestHandler.requiresInSampleSize(options);
 
     boolean isWebPFile = Utils.isWebPFile(stream);
+    boolean isPurgeable = request.purgeable && android.os.Build.VERSION.SDK_INT < 21;
```
sucks we still cant use the constant since we are not on gradle.

dnkoutso commented on picasso/src/main/java/com/squareup/picasso/Request.java:
```diff
@@ -79,6 +79,7 @@
   public final float rotationPivotY;
   /** Whether or not {@link #rotationPivotX} and {@link #rotationPivotY} are set. */
   public final boolean hasRotationPivot;
+  public final boolean purgeable;
```
nit: add a comment here.

dnkoutso commented:
LGTM. Want to address the minor nits and I will merge, have you signed CLA?

sneuberger-amazon commented:
Added the comment, and a couple minor fixes. I have signed the CLA.

ypresto commented:
Note that this feature should not be used in kitkat because there is nasty file descriptor leak bug in BitmapFactory...

https://code.google.com/p/android/issues/detail?id=65638#makechanges
